### PR TITLE
aggregated remaining patches to correct use of extern "C"

### LINKED ISSUE #17997: fix extern "C" use throughout Zephyr
Inspired by #17993, the latest in an unending series of patches to fix broken C++ support, I did:
```
lilith[632]$ git grep -A12 'extern "C"' | grep '#include' | sed -e 's@\.h-.*$@.h@' | uniq | wc
    303     303   15609
```
That's at least 303 include files in the Zephyr tree that probably get this wrong, i.e. have an `#include` within the first ten lines after an `extern "C"` block.  Excluding `ext/hal` (found in atmel, nordic, nxp, openisa, ti) reduces it to a mere 88.

As long as Zephyr claims C++ compatibility, even if only for application code, this should be cleaned up at least in non-HAL code before 2.0, and there should be some attempt toward adding a check for new occurrences to the commit checking code.

## Rationale 
Declarations that use C linkage should be placed within extern "C" so the language linkage is correct when the header is included by a C++ compiler.

Similarly `#include` directives should be outside the extern "C" to ensure the language-specific default linkage is applied to any declarations provided by the included header.

See: https://en.cppreference.com/w/cpp/language/language_linkage

**Note** Some ["authorities"](https://stackoverflow.com/questions/67894/) specifically *recommend* putting include files within `extern "C"`.  Ignore them.  That solution is fragile because the supplier of the header may expect that when included from C++ the default linkage is correct for C++.  This expectation is made by some Zephyr headers, which will produce compiler errors if included by a C++ compiler within an `extern "C"` block.  Place the burden of assigning declaration linkage on the header that provides the declaration, rather than specifying it in the including header/code.

### PR DESCRIPTION
**Pull request by pabigot**
This PR, in combination with #18243 and #18244, should be sufficient to close #17997.

If the complexity is too high it can be broken down into chunks.  I figured one big PR was better than the eight smaller ones that fell out of an initial partition.

### CODE REVIEW

**Review by tyrantlizard77 — APPROVED** [review]
What can we do to make this not a constant source of bit-rot?

**Comment by pabigot** [comment]
@andrewboie 

> What can we do to make this not a constant source of bit-rot?

Something like https://github.com/zephyrproject-rtos/zephyr/issues/17997#issuecomment-520615877 could be added to the CI scripts to at least indicate the presence of nested includes.  That technique is a bit fragile in the long term as `__cplusplus`  might be used for something other than linkage blocks.

**Review by ioannisg — APPROVED** [review]
I think this is a good cleanup patch, reviewers, pls take a look

**Review by carlescufi — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by wentongwu — APPROVED** [review]

File: arch/arc/include/kernel_arch_data.h [hunk 1]
```diff
@@ -36,9 +32,11 @@ extern "C" {
 #include <zephyr/types.h>
 #include <sys/util.h>
 #include <sys/dlist.h>
+
+#ifdef __cplusplus
+extern "C" {
 #endif
 
-#ifndef _ASMLANGUAGE
```
Comment (ruuddw) [inline arch/arc/include/kernel_arch_data.h]: other _ASMLANGUAGE checks in this commit have been moved, not removed. Are you sure this one can be removed?
The corresponding #endif in line 173/175 isn't removed.

**Review by ruuddw — CHANGES_REQUESTED** [review]
see comment for arch/arc/include/kernel_arch_data.h

File: arch/arc/include/kernel_arch_data.h (hunk 1 above)
Comment (pabigot) [reply arch/arc/include/kernel_arch_data.h]: What's actually happening here is that the `#endif` on line 38 used to be the close of the `#ifndef _ASMLANGUAGE` up on line 33 that enclosed only the `#include` material, immediately followed by another `#ifndef _ASMLANGUAGE` for the structure definition.    So the previous code had two adjacent `#ifndef _ASMLANGUAGE/#endif` blocks, which I merged by removing the `#ifdef/#endif` pair.   I then inserted an `#ifdef __cplusplus/#endif` block which restored the `#endif` on 39.  The whole `#ifndef _ASMLANGUAGE` block is still closed down on line 173.

See https://github.com/zephyrproject-rtos/zephyr/blob/0409dd1ce83501b2e091edb9ccbf70e92e89a967/arch/arc/include/kernel_arch_data.h#L33 for the before state.

OK?

Comment (ruuddw) [reply arch/arc/include/kernel_arch_data.h]: thanks for clarifying, missed that when only looking at the changes and not complete file.

**Review by ruuddw — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Pull request merged**